src/main.cpp

```diff
@@ -5170,6 +5170,13 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             return true;
         }
 
+        if (mapBlockIndex.find(headers[0].hashPrevBlock) == mapBlockIndex.end()) {
+            // Doesn't connect, instead of DoSing in AcceptBlockHeader, request deeper headers
+            if (!IsInitialBlockDownload())
+                pfrom->PushMessage(NetMsgType::GETHEADERS, chainActive.GetLocator(pindexBestHeader), uint256());
```

Comment:
Well my guess is that this isn't a theoretical problem, but an actual one -- I just started grepping the debug logs on one of my nodes, and I'm seeing roughly 1 message/day of the form "ERROR: AcceptBlockHeader: prev block not found".  I didn't check in depth, but I believe that corresponds to the condition that the handling of getheaders messages is broken, which would mean the infinite loop behavior would be triggered.

Not obvious to me what the best solution is.  We could add more per-node state, perhaps a single `bool fBadHeadersImplementation` that is set to false whenever a valid headers message is processed from a peer, and when processing a headers message that doesn't connect, set it to true and request more headers, unless it's already true, in which case we give up?  But I haven't tried to implement so not sure how tricky this would be...